### DISCUSSION TITLE
CAPI app stopping

### DISCUSSION THREAD

**Discussion by gigagrig** [uNetworking/uWebSockets#1521] [Q&A] [answered]
uWebSockets C++ library has way to safely stop server by scheduling socket closing with code like that.

`loopToShutdown->defer([]() {
  us_listen_socket_close(...)
});`
The code can be invoked from any thread.

Does CAPI has a similar or another way to stop the app?

**Comment by cirospaciari** [comment, marked as answer]
@gigagrig 

In my case i just use [libuv](https://docs.libuv.org/en/v1.x/api.html) to defer to me the close call, loop defer its not yet on C API.

You can get the native loop using:
```c
    struct us_loop_t *uws_get_loop();
    struct us_loop_t *uws_get_loop_with_native(void* existing_native_loop);
```
```uws_get_loop_with_native```basically sets you the event loop and returns to you

```c
    uv_loop_t *loop = (uv_loop_t *loop)uws_get_loop_with_native(uv_default_loop());
```
or

```c
    uv_loop_t *loop =  ( uv_loop_t *)uws_get_loop();
```

and after this you can use any API of [libuv](https://docs.libuv.org/en/v1.x/api.html) like [uv_prepare](https://docs.libuv.org/en/v1.x/prepare.html) to close the API before pooling more IO, or [uv_check](https://docs.libuv.org/en/v1.x/check.html) to close the API after pooling more IO, if you want to call this from another thread (remembers that you can only have 1 App/SSLApp per thread) you can use [async](https://docs.libuv.org/en/v1.x/async.html) to wakeup the event loop in main thread and close to you

but libuv is not the only loop option, so i think is a good idea to add defer to CAPI in future

**Reply by gigagrig** [reply]
@cirospaciari thanks for detailed answer! 
Now i'm starting to realize that using capi becomes more and more tricky.
So am a right that if i need to defer a reply, i also have to use API of libuv too? Because I cannot call `uws_res_end` and other reply functions in other than App/SSLApp thread?

**Reply by cirospaciari** [reply]
No you don't need to use libuv and you can't never use this library from other threads (it's not a thread safe library it's not the CAPI it's the design of uWS see the docs), you can use multiple threads (one app per thread) or signaling between threads, you don't need to use libuv, you can see examples on CAPI without libuv and using basically all resources.

**Reply by gigagrig** [reply]
I got it. I realized that I should use `us_timer_set` for async replies logic.
Thanks for help again!
